doc/computing/parallelism.rst

```diff
@@ -10,38 +10,42 @@ Parallelism, resource management, and configuration
 Parallelism
 -----------
 
-Some scikit-learn estimators and utilities can parallelize costly operations
-using multiple CPU cores, thanks to the following components:
+Some scikit-learn estimators and utilities parallelize costly operations
+using multiple CPU cores.
 
-- via the `joblib <https://joblib.readthedocs.io/en/latest/>`_ library. In
-  this case the number of threads or processes can be controlled with the
-  ``n_jobs`` parameter.
-- via OpenMP, used in C or Cython code.
+This can be either done:
 
-In addition, some of the numpy routines that are used internally by
-scikit-learn may also be parallelized if numpy is installed with specific
-numerical libraries such as MKL, OpenBLAS, or BLIS.
+- with higher-level parallelism via `joblib <https://joblib.readthedocs.io/en/latest/>`_.
+  In this case, the number of threads or processes can be controlled with the
+  ``n_jobs`` parameter.
+- with lower-level parallelism via OpenMP, used in C or Cython code.
+  In this case, parallelism is always done using threads and specifying
+  ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally
+  more performant than joblib-based implementations by up to two orders of magnitude.
```

Comment:
I would remove this last sentence because those "joblib-based implementations" are no longer available to the scikit-learn user. I think this can cause confusion for users who might wondering where they are and how they can be enabled.

The user guide should only refer to the current state of the library, not about the evolution between releases. The changelog and blog posts are better for that.

Also the main reason why the new Cython implementations are orders of magnitude faster than previous numpy-based implementations is primarily because they rely on fine-grained chunking to avoid RAM transfers temporary arrays and not necessarily because of joblib-managed threads vs OpenMP-managed threads. Those (important) implementation details are outside of the scope of this paragraph.

Reply:
Instead of stating:

"In this case, parallelism is always done using threads and specifying ``n_jobs`` *has no effect*"

one could add a new sentence as a new paragraph after the end of the 2 bullet point such as:

> The `n_jobs` parameters of estimators always controls the amount of parallelism managed by joblib (processes or threads depending on the joblib backend). The thread-level parallelism managed by OpenMP in scikit-learn's own Cython code or by BLAS/LAPACK libraries used by numpy and scipy operations used in scikit-learn is always controlled by environment variables or `threadpoolctl` as explained below. Note that some estimators can leverage all 3 kinds of parallelism at different points of their training and prediction methods.

Reply:
> I would remove this last sentence because those "joblib-based implementations" are no longer available to the scikit-learn user. I think this can cause confusion for users who might wondering where they are and how they can be enabled.

There's an option in the scikit-learn configuration to disable `PairwiseDistancesReduction` if needed.

> Also the main reason why the new Cython implementations are orders of magnitude faster than previous numpy-based implementations is primarily because they rely on fine-grained chunking to avoid RAM transfers temporary arrays and not necessarily because of joblib-managed threads vs OpenMP-managed threads. Those (important) implementation details are outside of the scope of this paragraph.

It is true that this is out of the scope of this section.